File: python/binary-tree-general/path_sum.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def hasPathSum(self,root: Optional[TreeNode], targetSum: int) -> bool:
        if not root:
            return False

        if not root.left and not root.right:
            return targetSum == root.val


        targetSum -= root.val

        return self.hasPathSum(root.left, targetSum) or self.hasPathSum(root.right, targetSum)

File: java/src/main/java/com/sanedge/interview/binary_tree_general/PathSum.java
package com.sanedge.interview.binary_tree_general;

import com.sanedge.interview.utils.TreeNode;

public class PathSum {

    public boolean hasPathSum(TreeNode root, int targetSum) {
        if (root == null) {
            return false;
        }

        if (root.left == null && root.right == null) {
            return targetSum == root.val;
        }

        targetSum -= root.val;

        return (
            hasPathSum(root.left, targetSum) ||
            hasPathSum(root.right, targetSum)
        );
    }
}
